Add flic smart button component

soldag opened the pull request:
**Description:**
This component let users use flic buttons to trigger automations.

**Related issue (if applicable):** fixes #<home-assistant issue number goes here>

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#1527

**Example entry for `configuration.yaml` (if applicable):**
```yaml
binary_sensor:
  - platform: flic
```

**Checklist:**

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [ ] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L16
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L51

soldag committed:
Added component for flic smart buttons.

soldag committed:
Apply home-assistant coding styles.

soldag committed:
Fixed flic configuration.

soldag committed:
Made logging for scanning for new buttons less verbose.

soldag committed:
Fixed flic event data.

soldag committed:
Follow async conventions.

soldag committed:
Added new requirements to requirements_all.txt.

soldag committed:
Added flic component to .coveragerc

soldag committed:
Updated flic threshold configuration key names.

soldag committed:
Flic devices are now removed when they disconnect.

balloob commented on homeassistant/components/binary_sensor/flic.py:
```diff
@@ -0,0 +1,240 @@
+
+    def new_button_callback(address):
+        """Setup newly verified button as device in home assistant."""
+        hass.loop.create_task(async_setup_button(hass, config,
```
Please use `hass.add_job`

balloob commented on homeassistant/components/binary_sensor/flic.py:
```diff
@@ -0,0 +1,240 @@
+    import pyflic
+
+    # Get event loop
+    loop = asyncio.get_event_loop()
```
Preferred to use

```
loop = hass.loop
```

balloob commented on homeassistant/components/binary_sensor/flic.py:
```diff
@@ -0,0 +1,240 @@
+        if not was_queued:
+            # Set state
+            self._is_down = click_type == pyflic.ClickType.ButtonDown
+            self.update_ha_state()
```
Please use `self.schedule_update_ha_state()`

balloob commented on homeassistant/components/binary_sensor/flic.py:
```diff
@@ -0,0 +1,240 @@
+
+            if self._is_down:
+                self._last_down = now
+            else:
```
Instead of this `else`, include a `return` in the `if` statement above and you don't have to indent the other code that much

balloob commented on homeassistant/components/binary_sensor/flic.py:
```diff
@@ -0,0 +1,240 @@
+        """Fire click event depending on click type."""
+        import pyflic
+
+        if not was_queued:
```
Prefer to use a guard clause

```python
if was_queued:
    return

# Set state
…
```

balloob commented on homeassistant/components/binary_sensor/flic.py:
```diff
@@ -0,0 +1,240 @@
+        if result == pyflic.ScanWizardResult.WizardSuccess:
+            _LOGGER.info("Found new button (%s)", address)
+        elif result != pyflic.ScanWizardResult.WizardFailedTimeout:
+            _LOGGER.info("Failed to connect to button (%s). Reason: %s",
```
Should this be `_LOGGER.warning` ?

balloob commented on homeassistant/components/binary_sensor/flic.py:
```diff
@@ -0,0 +1,240 @@
+    if auto_scan:
+        scan_client = pyflic.FlicClient(host, port)
+        start_scanning(hass, config, async_add_entities, scan_client)
+        loop.run_in_executor(None, scan_client.handle_events)
```
Will this call ever finish ? Or will it occupy an executor thread forever?

balloob commented on homeassistant/components/binary_sensor/flic.py:
```diff
@@ -0,0 +1,240 @@
+
+    main_client.on_new_verified_button = new_button_callback
+    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, main_client.close)
+    loop.run_in_executor(None, main_client.handle_events)
```
Will this call ever finish ? Or will it occupy an executor thread forever ?

soldag replied:
No, `FlicClient.handle_events()` has to run the entire time in background. It is a main loop of the pyflic lib that never exits, unless the socket to the flic service is closed.

balloob reviewed (commented):
Alright, so this looks pretty good.

I couldn't follow all the code. It looks like it is callback based so that's good. Please ensure that you're not doing any I/O inside the a coroutine.

soldag committed:
Include review feedback.

soldag commented:
@balloob Thanks :)

Yes, the library is callback based. The only method that is doing I/O is `FlicClient.handle_events()`as already mentioned in the comments.

soldag committed:
Fixed stopping of clients in flic component when home assistant is stopped.

armills commented:
Related question here, mostly for @balloob or other maintainers. Does it make sense for hass to have a stateless button component instead of using the binary_sensor component? In this case we're setting the state based on when the button is down, but it's only a momentary change. The useful feature exposed by this platform is the click events. Looking forward to other platforms like Amazon dash, we may not get this state information, and only get a single trigger.

soldag committed:
Updated flic component by integrating input of #4738.

Use library method to determine click type. Merge three click events into single one with click_type parameter.

soldag commented:
@armills Thanks for your work, even if I was a little bit faster 😜 
Anyway, I had a look at your commit and updated my component. I didn't realize that there is a `on_button_single_or_double_click_or_hold` method in fliclib and had built it myself, but thanks to you it is now used. :satisfied:

armills commented:
@soldag Looks great! Yeah, they don't show it in their examples. I just stumbled across it when looking through the library. At least it wasn't for nothing!

Can you take a look at my other two notes? They're just some smaller questions I had.

soldag commented:
@armills Where did you take this notes? I couldn't find them.

armills commented:
They should show up when you scroll through here: https://github.com/home-assistant/home-assistant/pull/4681/files/8adb707e3262635f5b77530defea5d6cbbe26afa

soldag commented:
Yeah, I thought so, but there is nothing but the code.

armills commented:
OK, thanks Github. I'll just put them here.

1) Instead of CONF_AUTO_SCAN, does it make sense to use CONF_DISCOVERY from homeassistant.const, just so we're using existing conventions?

2) [In the auto scan logic](https://github.com/home-assistant/home-assistant/pull/4681/files#diff-afed563a4ae6b8016c542f1a481cd4f0R62), can you just pass the `main_client` to `start_scanning` instead of creating a new client?

soldag committed:
Use a single client for both handling click events and scanning for new buttons.

soldag committed:
Renamed flic ‘auto_scan’ configuration variable to ‘discovery’ using HA constants.

soldag commented:
These suggestions make totally sense. I updated the code 😉

balloob commented:
@armills I think that a button component would be overkill here as there is no generic interface to abstract out, we're just firing events. Things like this should probably just be their own component (we can still categorize them as button on the HA site to help people discover them). Then again, I don't mind it being a binary sensor either.

For now I will merge this PR as it all works.

armills commented:
@balloob OK, makes sense. Thanks for the reply!

fabianbergmark commented:
You should also use the other click callbacks, i.e. on_button_click_or_hold and on_button_single_or_double_click. Also, queued events might be of interest to some users so I suggest not discarding them and instead include the time_diff in the event data.

soldag commented:
@fabianbergmark What's the use of the other click callbacks? I thought the others are used to only listen to specific events and as I want to listen to all types of click events, `on_button_single_or_double_click_or_hold` is the only one I need. 

The point about queued events makes sense and I will update the component code.

fabianbergmark commented:
The logic behind the click callbacks is pretty smart, but not very well explained. Say you are not interested in double clicks but only click and hold. If you click the button twice fast, the on_button_click_or_hold would fire twice with a click event, while the on_button_single_or_double_click_or_hold would only fire once with a double click event. This is very useful for applications where you want to click fast but still have a secondary hold trigger.

armills commented:
I opened a PR to support queued events here: https://github.com/home-assistant/home-assistant/pull/4822

soldag commented:
@fabianbergmark Okay, now I understand 👍 
I am now making use of this by giving the user the possibility to ignore individual click types. Depending on this selection the appropriate callback is registered. See PR #4827 .

AntonMeier commented:
One last note on the click topic is that if you discard the double click event then you will get lower latency on the click event since the "logic" does not have to wait for a possible second click in a double click sequence.

Emill commented:
Hi. I don't think it should be default to constantly scan in the background.
That means someone can stand outside the window and easily connect a button into your home automation system...
A less important point is that scanning takes up more radio time which could increase latency for other Flics as well as lowering the WiFi throughput, as well as consuming more power.

armills commented:
[You can disable the scan by setting `discovery: false` in your config.](https://home-assistant.io/components/binary_sensor.flic/#configuration) It is enabled by default because at least one button needs to be discovered.

As far as security concerns go, barring an unknown vulnerability in flicd, the worst an attacker could do is a DOS, which they can probably do anyway by saturating the Bluetooth frequency.

The scanning for new buttons is passive. The daemon is listening for advertisement packets from the buttons. It isn't actively polling the network, so there shouldn't be an impact on radio usage. The only change when turning off discovery is that the daemon ignores advertisements from unknown buttons.

Emill commented:
Yes it can be disabled, but I don't want it to be default true since most users will just leave with the default settings. I would rather have some default mode which scans only while there is currently no button paired with the daemon.

I don't know if you got my point with the security problem. If you always run the scan wizard that means that anyone can stand outside the window and pair Flic buttons so they are now trusted by the daemon without the user noticing anything. If you open the Bluetooth settings on your smartphone and you spontaneously get a popup asking if it's OK to pair with some random device you've never before, you usually press NO, right?

If you run a Raspberry Pi 3, then the WiFi chip and Bluetooth chip is the same thus it also shares the same radio, which can only do one thing at a time. When you run a Bluetooth scan, you reserve radio usage in certain time frames which blocks out all other traffic during that time window. Probably most users won't notice anything but I did that test once on a smartphone and the network speed indeed became slower when a Bluetooth scan was active.

Emill commented:
You must also check the status code of when the wizard completes. If the bluetooth dongle is currently unplugged, it will complete immediately with status WizardBluetoothUnavailable. With this current implementation that leads to an infinite busy loop. Instead, if you get WizardBluetoothUnavailable you should not try again but rather upon receiving EvtBluetoothControllerStateChange with state = Attached, then it is ok to try again.

armills commented:
That's a good point. Can you open two issues on GitHub for these comments so they can be discussed in a clean thread? You could also just open a PR for the bluetooth unavailable if you want to write it up and test it.
